bibtex-actions-filenotify.el

```diff
@@ -0,0 +1,159 @@
+;;; bibtex-actions-filenotify.el --- filenotify functions for bibtex-actions -*- lexical-binding: t; -*-
+;;
+;; Copyright (C) 2021 Bruce D'Arcus
+;;
+;; Author: Bruce D'Arcus <https://github.com/bdarcus>
+;; Maintainer: Bruce D'Arcus <bdarcus@gmail.com>
+;; Created: August 17, 2021
+;; Modified: August 17, 2021
+;; Version: 0.0.1
+;; Keywords: abbrev bib c calendar comm convenience data docs emulations extensions faces files frames games hardware help hypermedia i18n internal languages lisp local maint mail matching mouse multimedia news outlines processes terminals tex tools unix vc wp
+;; Homepage: https://github.com/bruce/bibtex-actions-filenotify
+;; Package-Requires: ((emacs "24.3"))
+;;
+;; This file is not part of GNU Emacs.
+;;
+;;; Commentary:
+;;
+;;
+;;
+;;; Code:
+
+(require 'filenotify)
+(require 'files)
+
+(declare-function bibtex-actions-refresh "bibtex-actions")
+(declare-function bibtex-actions--local-files-to-cache "bibtex-actions")
+(declare-function bibtex-actions--global-files-to-cache "bibtex-actions")
+(declare-function bibtex-actions-utils--normalize-paths "bibtex-actions-utils")
+(declare-function reftex-access-scan-info "ext:reftex")
+(declare-function file-notify-add-watch "ext:file-notify")
+(declare-function file-notify-rm-watch "ext:file-notify")
+
+(defvar bibtex-actions-bibliography)
+(defvar bibtex-actions-library-paths)
+(defvar bibtex-actions-notes-paths)
+(defvar bibtex-actions-create-note-function)
+(defvar bibtex-actions-filenotify-files)
+(defvar bibtex-actions--candidates-cache)
+(defvar bibtex-actions--local-candidates-cache)
+(defvar bibtex-actions-filenotify-callback)
+(defvar bibtex-actions-open-file-function)
+(defvar bibtex-actions-file-extensions)
+
+;;; File watching
+
+(defvar-local bibtex-actions--local-watches 'uninitialized)
+(defvar bibtex-actions--global-watches nil)
+
+(defun bibtex-actions--invalidate-cache (&optional scope)
+  "Invalidate local or global caches according to SCOPE.
+If it is other than 'global or 'local invalidate both"
+  (unless (eq 'local scope)
+    (setq bibtex-actions--candidates-cache 'uninitialized))
+  (unless (eq 'glocal scope)
+    (setq bibtex-actions--local-candidates-cache 'uninitialized)))
+
+(defun bibtex-actions--make-default-callback (func scope &optional change)
+  "The callback FUNC by SCOPE used to update cache for default options.
```

Comment:
@aikrahguzar - in which file should these go?

Reply:
`bibtex-actions--make-default-callback` should be in `bibtex-actions-filenotify`. It is pretty much an  internal detail. 

I don't have a strong opinion on `bibtex-actions--invalidate-cache`. I don't  see a reason to use it over `bibtex-actions-refresh` for anything other the filenotify callback so I think it can be moved to `bibtex-actions-filenotify` too.

Reply:
I think even moving two `defcustoms` and `bibtex-actions-filenotify-setup` as the only autload of the file is an option. The only minor downside I see is that it will require a separate use-package call to configure it in that case. On the positive side it will make the `filenotify` stuff self-contained and we can remove the duplicate `defvar`s from `bibtex-actions` in `bibtex-actions-filenotify`, remove `(require filenotify)` in `bibtex-actions` and instead do `(require bibtex-actions)` in `bibtex-actions-filenotify` which seems a more logical dependency.

Reply:
> I think even moving two `defcustoms` and `bibtex-actions-filenotify-setup` as the only autload of the file is an option. The only minor downside I see is that it will require a separate use-package call to configure it in that case. On the positive side it will make the `filenotify` stuff self-contained and we can remove the duplicate `defvar`s from `bibtex-actions` in `bibtex-actions-filenotify`, remove `(require filenotify)` in `bibtex-actions` and instead do `(require bibtex-actions)` in `bibtex-actions-filenotify` which seems a more logical dependency.

I just did this in the latest commit.

Aside from the question I have below, I think this is ready for a look, if and when you have time.

I think this is much cleaner.

Reply:
> Aside from the question I have below, I think this is ready for a look, if and when you have time.

I don't see any question?